### PR TITLE
Add service unavailability detection.

### PR DESCRIPTION
**Pull request by tasdomas**

### CODE REVIEW

File: api/budget/api.go [hunk 1]
```diff
@@ -173,6 +174,9 @@ func (c *client) doRequest(req interface{}, result interface{}) error {
 		}
 		resp, err = c.h.DoWithBody(req, bytes.NewReader(payload.Bytes()))
 		if err != nil {
+			if strings.HasSuffix(err.Error(), "Connection refused") {
```
Comment (cmars) [inline api/budget/api.go]: Is there a stronger test for this (perhaps by using juju/errors or errgo, whatever the return error is wrapped with), or is the underlying error completely masked/flattened?

File: api/budget/api.go [hunk 2]
```diff
@@ -173,6 +174,9 @@ func (c *client) doRequest(req interface{}, result interface{}) error {
...
+				return wireformat.NotAvailError{}
+			}
 			return errors.Annotate(err, "failed to execute request")
```
Comment (cmars) [inline api/budget/api.go]: This might be good enough to troubleshoot a connection failure with. It would certainly differentiate it from a 503.

**Comment by cmars** [comment]
LGTM. I'm a little concerned about parsing the error message for connection failure -- perhaps that's not necessary?

**Comment by tasdomas** [comment]
Could not find a better way of detecting a connection failure. errors.Annotate is not enough, because in https://github.com/juju/juju/pull/4280 I need to return nil in case I detect a connection error.

**Comment by cmars** [comment]
Makes sense. Still LGTM.

**Comment by tasdomas** [comment]
@cmars, I don't have write access to this repo, could you land this for me?

**Pull request merged**